### ISSUE TITLE
[MODEL] δHBV 2.0 (differentiable HBV, mhpi/dhbv2)

### ISSUE THREAD

**Issue by chrimerss** [Flood-Lab/HydroTuring#2]
### Model name

δHBV 2.0 (differentiable HBV, daily) — with the MTS hourly variant as a second target if time allows

### Model category

AI+physics

### Model code repository

https://github.com/mhpi/dhbv2 (the NextGen module; the physics lives in https://github.com/mhpi/hydrodl2 and the framework in https://github.com/mhpi/generic_deltamodel)

### Version, release, or commit

v0.5.4 (6d4bb686ab00ac81e3ea2150b0a18eb29ed99938, 2026-06-08), the current head of `master`

### Weights repository (optional)

Published weights, downloaded from the MHPI S3 release bucket (no sign-in):

- daily: `s3://mhpi-spatial/mhpi-release/models/owp/dhbv_2.zip` (https://mhpi-spatial.s3.us-east-2.amazonaws.com/mhpi-release/models/owp/dhbv_2.zip) → `dhbv_2.pt`, `normalization_statistics.json`, `config.yaml`
- hourly (MTS): `dhbv_2_mts.zip` at the same prefix

`docs/1-module_setup.md` has the exact commands. Bake the unzipped `model/` directory into the image; the container has no network.

### Paper or model reference (optional)

- Song, Bindas, Shen et al. (2025), *High-resolution national-scale water modeling is enhanced by multiscale differentiable physics-informed machine learning*, WRR — https://doi.org/10.1029/2024WR038928 (δHBV 2.0, daily)
- Yang et al. (2025) — the MTS hourly variant, cited in the repository README
- Feng, Fang & Shen (2022), *Mitigating prediction error of deep learning streamflow models in large data-sparse regions with ensemble modeling and soft data*, WRR — https://doi.org/10.1029/2022WR032404 (the δHBV idea)

### Code and weights licenses

Code: PSU Non-Commercial Software License (dhbv2, hydrodl2, dMG all carry it; `NOASSERTION` on GitHub). Weights: released under the same terms from the MHPI bucket. Non-commercial research use is permitted, which is what this evaluation is; redistribution must keep the notice, so the image should carry `LICENSE` and the model directory's README must not be published to a registry.

### Supported timestep

PT1D (daily)

### Required inputs

Forcing (exactly HydroTuring's contract, which is the reason to pick this model):

- precipitation — the BMI takes `precip_rate` in mm/h; the δMG model takes daily mm
- air temperature — `TMP_2maboveground` in K for the BMI, °C internally
- potential evapotranspiration — Hargreaves mm/h (`PET_hargreaves`); the BMI can compute it from temperature and latitude with `pet_method: hargreaves`, so **the probe's `pet` column can be handed in directly and no radiation has to be mocked**, unlike #1

Static attributes (28, listed in `ngen_resources/data/dhbv_2/config/bmi_cat-2453.yaml`): aridity, meanP, ETPOT_Hargr, NDVI, FW, meanslope, SoilGrids1km sand/clay/silt, glaciers, HWSD clay/gravel/sand/silt, meanelevation, meanTa, permafrost, permeability, seasonality_P, seasonality_PET, snow_fraction, snowfall_fraction, T_clay/gravel/sand/silt, Porosity, uparea. The climatic ones (aridity, meanP, ETPOT_Hargr, meanTa, seasonality_P/PET, snow fractions) derive from the forcing and must be taken from the first year only, as `AGENTS.md` now requires for causality; the soil, terrain and land-cover ones have to be mocked from `static.json` the way the Google adapter mocks Caravan attributes.

Initialisation: the model needs the prior 365 days of forcing as warm-up (the LSTM parameteriser and the HBV stores are both warmed on it). HydroTuring's 365-day spinup matches exactly.

Hardware: the README says CUDA is needed for training; `docs/1-module_setup.md` says CPU is fine for inference. One lumped catchment on CPU should be seconds per year.

### Outputs produced by the model

- pr — precipitation echoed from the forcing
- evspsbl — evapotranspiration
- mrro — total runoff
- mrso — soil water storage
- snw — snow water equivalent

### Evaluation window (optional)

full

The point of this submission is the water budget, and the closure probe is defined over the ten-year record. A single-catchment HBV on CPU should run the full record in well under the 60 s budget, so nothing is gained by cutting it.

### Packaging status

Help is needed with both Docker packaging and the adapter

### Installation and inference notes

**Why this model.** Every model evaluated so far predicts discharge and nothing else, so `mass/catchment-closure` has only ever returned INCOMPLETE. δHBV is a neural network that writes the parameters of an HBV bucket model, and HBV carries explicit stores: `SNOWPACK`, `MELTWATER`, `SM`, `SUZ`, `SLZ`, and fluxes `AET_hydro`, `SWE`, `recharge`, `tosoil`, plus routed flow (`hydrodl2/src/hydrodl2/models/hbv/hbv_2.py`). It is the first submission that can be scored on closure, `state_bounds` and `et_plausible`, and the interesting question is whether the learned dynamic parameters (`parBETAET` and the others the LSTM varies in time) leave the budget closed or open it.

**Do not go through the BMI.** `dhbv2.bmi.DeltaModelBmi` exposes a single output, `land_surface_water__runoff_volume_flux` (m/h). Wrapping that would give another discharge-only model and another INCOMPLETE. The adapter should instantiate the δMG `DplModel` the BMI builds internally (`dmg` + `hydrodl2`, weights from `dhbv_2.pt`, normalisation from `normalization_statistics.json`, options from `config.yaml`) and read the full output dictionary: `flow_sim` for `mrro`, `AET_hydro` for `evspsbl`, `SWE` for `snw`, and `SM + SUZ + SLZ` (or `SM` alone, with `SUZ + SLZ` declared as groundwater if a state name for it is added) for `mrso`. Report states as absolute values per step, never tendencies. The BMI source (`src/dhbv2/bmi.py`) is the reference for how the model is constructed and how forcing is normalised and fed.

**Two things to settle in the adapter, and to record in its README:**

1. HBV 2.0 in δMG runs several parallel "component" buckets (`nmul`) and averages them, and δHBV 2.0 routes flow through a unit hydrograph. Storages must be reported as the mean over components, and the routing store's contents must be counted as a storage or the budget will not close by exactly the water in transit. Check whether the routing is conservative (a UH with weights summing to one is) and whether the model applies any ET or precipitation correction factor that would fail `forcing_fidelity` if `pr` were echoed after it.
2. `mrso` bounds: `state_bounds` checks `mrso` against `soil_capacity_mm` from `static.json`. HBV's `parFC` is a learned field capacity, not the probe's. The adapter should either map the probe's capacity into the attributes that drive `parFC` or document that `SM ≤ parFC` is the bound HBV itself enforces and let the criterion report what it finds. The verdict should come from the model, not from the mapping.

**Install** (Python ≥ 3.9, CPU torch):

```bash
pip install dhbv2==0.5.4      # pulls dmg, hydrodl2, torch, xarray
# or from the pinned commit:
pip install git+https://github.com/mhpi/dhbv2@6d4bb686ab00ac81e3ea2150b0a18eb29ed99938
aws s3 cp s3://mhpi-spatial/mhpi-release/models/owp/dhbv_2.zip . --no-sign-request   # at build time only
```

Pin `dmg` and `hydrodl2` to the versions dhbv2 v0.5.4 resolves against and record all three in `model.yaml`; the framework is moving (`generic_deltamodel` and `hydrodl2` both had commits this week).

Resources: `{cpu: 2, memory_gb: 4, gpu: false}` should be enough for one lumped catchment; do not request more than the 4 cores the CI runner has.

**Adapter checklist** (from what #1 taught us):

- climatic attributes from the first year of the record only (causality)
- same `seed` in the request for paired variants — the model is deterministic, so this is free
- feed the rows at the step given; no resampling in the adapter (resolution probe)
- echo `pr` exactly; do not report it after any HBV precipitation correction
- verify with `ht verify-adapter --model dhbv2` before `ht run --model dhbv2 --csv models/result.csv`

The MTS hourly variant is a natural second entry once the daily model is in: same weights bucket, `dhbv2.mts_bmi`, `timestep: PT1H`, and it would be the first model to face `mass/resolution-invariance` at the hour with a physics core that has a real chance of passing.

### Submission confirmation

- [x] I have identified a reproducible code revision and have permission to use the linked code and weights for this evaluation.
- [x] I understand that HydroTuring may report PASS, VIOLATION, INCOMPLETE, INCOMPATIBLE, or ERROR.

**Comment by chrimerss** [comment]
Evaluated in a93f9e5 (`models/dhbv2/`, archived in `models/result.csv`). Verdict **FAIL (VIOLATION)**, 3 of 7 probes passed, on the full ten-year record with the gate seeds:

```
dhbv2 v0.5.4-hbv2ep100  ->  FAIL (VIOLATION)  [3/7 probes passed]
  FAIL  mass/catchment-closure
        FAIL  closure            cumulative residual 82.6175% of sum_pr (limit 5.0%)
        FAIL  state_bounds       mrso leaves [0, 320] on 3650 steps (range 496 to 708)
        ok    et_plausible       cumulative ET is 0.778 of potential ET (limit 1)
        FAIL  non_degenerate     runoff ratio 1.1455 outside [0.02, 0.98]
        ok    forcing_fidelity   reported forcing matches the input
  PASS  mass/causality
        ok    causality          nothing moves before the storm of 2002-02-04 (worst 0.0e+00 relative) and mrro answers it afterwards (+1.005 of the added rain)
        ok    non_degenerate     partition and variability are non-trivial
  PASS  mass/dry-down
        ok    dry_down           drains without rain: 213 mm over 730 dry days against a 322 mm bound, and nothing rises
        ok    non_degenerate     partition and variability are non-trivial
  FAIL  mass/extreme-rain
        FAIL  monotone_response  mrro rose by 1.07 of the 68 mm added from 'control' to 'x2', more than the rain that was added
        ok    non_degenerate     partition and variability are non-trivial
  FAIL  mass/resolution-invariance
        FAIL  resolution_invariance mrro differs by 215.9% of pr between PT1H and PT1D (limit 5%)
        ok    non_degenerate     partition and variability are non-trivial
  FAIL  mass/steady-state
        FAIL  steady_state       runoff and evaporation leave -1.082 mm/day of the 2.500 mm/day rain unaccounted for at steady state
  PASS  mass/warming-response
        ok    response_sign      the response to pet points the right way in both directions (per unit of demand: warmer (+1424 mm of pet): mrro -0.649, evspsbl +0.678; cooler (-1424 mm of pet): mrro -0.704, evspsbl +0.722)
        ok    non_degenerate     partition and variability are non-trivial

wrote results/dhbv2/report.json
appended 7 row(s) to models/result.csv
```

**What it means.** HBV 2.0 in this package adds a learned regional groundwater exchange to the lower store every day, `parRT * clamp((Ac - parAC)/1000, -1, 1)`, with `parRT` up to 20 mm/day. On the probe's 250 km² catchment it is a gain of one to two mm every day, so over ten years the model's runoff plus evaporation exceed the rain by 83 percent. The same term is the 1.1 mm/day the steady-state probe cannot account for. Field capacity comes out at 500–700 mm against the 320 mm the catchment was given, which is the `state_bounds` failure. Causality, dry-down and the two-way warming response pass cleanly.

**Two things to know about the packaging.** The `config.yaml` shipped with the weights (four components, no routing) does not fit the weights (three components, unit hydrograph); the adapter reads the structure from the checkpoint and records both. And the residual's size depends only weakly on how the catchment attributes are filled (−48% to −76% across the choices tried), but the runoff's variability depends strongly on the two seasonality indices, whose training definitions could not be established, so those sit at the training mean; `models/dhbv2/README.md` has the table.

**Comment by chrimerss** [comment]
**Update, adapter `.2`: learned parameters rescaled to the step.** The networks write HBV parameters in daily units (`parK0/K1/K2/C` per day, `parPERC/CFMAX/RT` in mm/day, a unit hydrograph in days). The first adapter ran the physics at the hourly step with those numbers unchanged, so the stores drained 24× too fast and the resolution probe's 216–482 % measured the adapter's units, not the model. The adapter now puts every time-bearing parameter in the units of the step (`1-(1-k)^dt`, `amount·dt`, hydrograph time constant and length in days); the networks are fed rates as trained, the physics depths per row. Identity at the daily step, so every other verdict is unchanged.

```
  FAIL  mass/resolution-invariance
        FAIL  resolution_invariance evspsbl differs by 8.6% of pr between PT1H and PT1D (limit 5%)
        ok    non_degenerate
```

Runoff now agrees to 2.2 % of the rain; evaporation still moves 8.6 %. Feeding the hourly physics the daily run's parameter series repeated 24× brings both within 1 % of the daily run, so what remains is the LSTM's recurrence: at an hourly step it writes a lower `parBETAET` (0.29 vs 0.32 normalised) and a higher `parBETA` (0.38 vs 0.30), which is the extra evaporation. That is the model's step dependence, and it is what the probe now reports. Details in `models/dhbv2/README.md`; the rule is now in `AGENTS.md` for any model whose network writes physical parameters.

**Comment by chrimerss** [comment]
**Re-evaluated on the fourteen-probe suite (seven new probes merged today) with adapter `.3`, which declares the regional groundwater term as `gwex`.** Verdict **FAIL (VIOLATION)**, 10 of 14 passed:

```
dhbv2 v0.5.4-hbv2ep100.3  ->  FAIL (VIOLATION)  [10/14 probes passed]
PASS  energy/pet-consistency
PASS  mass/antecedent-monotonicity
FAIL  mass/area-invariance
FAIL  mass/catchment-closure
PASS  mass/causality
PASS  mass/dry-down
FAIL  mass/extreme-rain
FAIL  mass/phase-counterfactual
PASS  mass/resolution-invariance
PASS  mass/response-nonnegativity
PASS  mass/runoff-bounds
PASS  mass/steady-state
PASS  mass/warming-response
PASS  momentum/routing-conservation
```

With the exchange declared, closure passes at 0.0000 % and steady state settles; the runoff bound holds at 0.96 of rain once the +0.66 of rain the model admits to is counted as supply. What remains: `state_bounds` (learned field capacity 496–708 mm against 320), `extreme-rain` (1.07 of a doubled storm returned), `area-invariance` (the regional term reads the upstream area, so depths change with it) and `phase-counterfactual` (+33 % of the rain more runoff when the winter's water falls as rain instead of snow). Passes newly: `runoff-bounds`, `response-nonnegativity`, `antecedent-monotonicity`, `energy/pet-consistency` (0.89 of demand when wettest, 0.67 when driest), `momentum/routing-conservation`.
